pkg/crawl/crawler.go

```diff
@@ -60,55 +59,87 @@ type ProcessFunc func(ctx context.Context, path string, result Finding, version
 // the directory) will be ignored.
 func (c Crawler) Crawl(ctx context.Context, root string, match MatchFunc, process ProcessFunc) (Stats, error) {
 	stats := Stats{}
-	err := filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
-		if err != nil {
-			switch {
-			case os.IsPermission(err):
-				stats.PermissionDeniedCount++
-				return nil
-			case os.IsNotExist(err):
-				// Root should always exist, to pick up on misconfigured service, but log4j-sniffer can encounter transient
-				// files when walking, where WalkDir lists the directory entries but the entry disappears before or
-				// during the walk function iterating over it.
-				if path == root {
-					return err
-				}
-				return nil
-			}
-			return err
-		}
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		default:
+	rootFile, err := os.Lstat(root)
+	if err != nil {
+		return stats, err
+	}
+	if !rootFile.IsDir() {
+		err = c.processFile(ctx, &stats, root, rootFile.Name(), match, process)
+	} else {
+		err = c.processDir(ctx, &stats, root, match, process)
+	}
+	return stats, err
+}
+
+func (c Crawler) processDir(ctx context.Context, stats *Stats, path string, match MatchFunc, process ProcessFunc) error {
+	dirInfo, err := os.Open(path)
```

Comment:
This file isn't being closed.

We are going to want to use a named return value and deferred error check in the same way we do for `Identify`
https://github.com/palantir/log4j-sniffer/blob/8eb0f71cc89fa69943f92dfbdb5e81e6870d4a0c/pkg/crawl/identify.go#L114
https://github.com/palantir/log4j-sniffer/blob/8eb0f71cc89fa69943f92dfbdb5e81e6870d4a0c/pkg/crawl/identify.go#L150-L154